File: Python/0079.py
from typing import List
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:

        visited = [[False] * len(board[0]) for _ in range(len(board))]

        def dfs(i, j, idx):
            if idx == len(word):
                return True
            
            if i < 0 or i >= len(board) or j < 0 or j >= len(board[0]) or visited[i][j] or board[i][j] != word[idx]:
                return False

            visited[i][j] = True

            if dfs(i+1, j, idx+1) or dfs(i-1, j, idx+1) or dfs(i, j+1, idx+1) or dfs(i, j-1, idx+1):
                return True
            
            visited[i][j] = False
            
            return False
        
        for i in range(len(board)):
            for j in range(len(board[0])):
                if dfs(i, j, 0):
                    return True
        return False 
                    
# Topic: Backtracking, DFS
# Time Complexity: O(N * 4^L) N is the number of cells in the board and L is the length of the word
# Space Complexity: O(N) N is the number of cells in the board

File: C++/0079.cpp
#include <vector>
#include <string>

class Solution {
public:
    bool dfs(int i, int j, int idx, std::vector<std::vector<char>>& board, std::string word ){
        if(word.length() == idx) return true;
        if(i < 0 || i >= board.size() || j < 0 || j >= board[0].size() || board[i][j] != word[idx]) return false;
        char temp = board[i][j];
        board[i][j] = ' ';

        bool ans = dfs(i+1, j, idx+1, board, word) || dfs(i-1, j, idx+1, board, word) || dfs(i, j+1, idx+1, board, word) || dfs(i, j-1, idx+1, board, word);
        
        board[i][j] = temp;
        return ans;
    }

    bool exist(std::vector<std::vector<char>>& board, std::string word) {
        for (int i = 0; i < board.size(); i++){
            for (int j = 0; j < board[0].size(); j++){
                if(dfs(i, j, 0, board, word)){
                    return true;
                }
            }
        }
        return false;
    }
};

// Topic: Array, Backtracking
// Time Complexity: O(N*M*4^L)
// Space Complexity: O(L)